File: apps/frontend/src/engine/state/ringBuffer.ts
/**
 * Generic ring buffer implementation with O(1) push/pop operations
 */
export class RingBuffer<T> {
  private buffer: (T | undefined)[];
  private head = 0;
  private tail = 0;
  private count = 0;
  private readonly capacity: number;

  constructor(capacity: number) {
    if (capacity <= 0) {
      throw new Error('Ring buffer capacity must be positive');
    }
    this.capacity = capacity;
    this.buffer = new Array(capacity);
  }

  /**
   * Add an item to the buffer
   * If buffer is full, removes the oldest item
   */
  push(item: T): void {
    this.buffer[this.tail] = item;
    this.tail = (this.tail + 1) % this.capacity;
    
    if (this.count < this.capacity) {
      this.count++;
    } else {
      // Buffer is full, move head to overwrite oldest item
      this.head = (this.head + 1) % this.capacity;
    }
  }

  /**
   * Remove and return the oldest item
   * Returns undefined if buffer is empty
   */
  pop(): T | undefined {
    if (this.count === 0) {
      return undefined;
    }

    const item = this.buffer[this.head];
    this.buffer[this.head] = undefined;
    this.head = (this.head + 1) % this.capacity;
    this.count--;
    
    return item;
  }

  /**
   * Get the most recent item without removing it
   */
  peek(): T | undefined {
    if (this.count === 0) {
      return undefined;
    }
    
    const lastIndex = (this.tail - 1 + this.capacity) % this.capacity;
    return this.buffer[lastIndex];
  }

  /**
   * Get the oldest item without removing it
   */
  peekOldest(): T | undefined {
    if (this.count === 0) {
      return undefined;
    }
    
    return this.buffer[this.head];
  }

  /**
   * Get the number of items currently in the buffer
   */
  size(): number {
    return this.count;
  }

  /**
   * Check if the buffer is empty
   */
  isEmpty(): boolean {
    return this.count === 0;
  }

  /**
   * Check if the buffer is full
   */
  isFull(): boolean {
    return this.count === this.capacity;
  }

  /**
   * Get the maximum capacity of the buffer
   */
  getCapacity(): number {
    return this.capacity;
  }

  /**
   * Clear all items from the buffer
   */
  clear(): void {
    this.buffer.fill(undefined);
    this.head = 0;
    this.tail = 0;
    this.count = 0;
  }

  /**
   * Get all items in order (oldest to newest)
   */
  toArray(): T[] {
    if (this.count === 0) {
      return [];
    }

    const result: T[] = [];
    for (let i = 0; i < this.count; i++) {
      const index = (this.head + i) % this.capacity;
      const item = this.buffer[index];
      if (item !== undefined) {
        result.push(item);
      }
    }
    return result;
  }

  /**
   * Get the last N items in order (oldest to newest)
   */
  getLast(n: number): T[] {
    const allItems = this.toArray();
    return allItems.slice(-n);
  }

  /**
   * Iterate over all items in order (oldest to newest)
   */
  *[Symbol.iterator](): Iterator<T> {
    for (let i = 0; i < this.count; i++) {
      const index = (this.head + i) % this.capacity;
      const item = this.buffer[index];
      if (item !== undefined) {
        yield item;
      }
    }
  }

  /**
   * Iterate over the last N items in order (oldest to newest)
   */
  *iterateLast(n: number): Iterator<T> {
    const items = this.getLast(n);
    for (const item of items) {
      yield item;
    }
  }

  /**
   * Find the first item that matches the predicate
   */
  find(predicate: (item: T) => boolean): T | undefined {
    for (const item of this) {
      if (predicate(item)) {
        return item;
      }
    }
    return undefined;
  }

  /**
   * Find the last item that matches the predicate
   */
  findLast(predicate: (item: T) => boolean): T | undefined {
    const items = this.toArray();
    for (let i = items.length - 1; i >= 0; i--) {
      if (predicate(items[i])) {
        return items[i];
      }
    }
    return undefined;
  }

  /**
   * Filter items that match the predicate
   */
  filter(predicate: (item: T) => boolean): T[] {
    const result: T[] = [];
    for (const item of this) {
      if (predicate(item)) {
        result.push(item);
      }
    }
    return result;
  }

  /**
   * Map items to a new type
   */
  map<U>(mapper: (item: T) => U): U[] {
    const result: U[] = [];
    for (const item of this) {
      result.push(mapper(item));
    }
    return result;
  }
}

/**
 * Create a ring buffer with the specified capacity
 */
export const createRingBuffer = <T>(capacity: number): RingBuffer<T> => {
  return new RingBuffer<T>(capacity);
};

File: cloudshell-temp/backend/indicator_engine/core/ring_buffer.py
"""Ring buffer implementation for efficient data storage."""

from typing import List, Optional, TypeVar, Generic
import threading

T = TypeVar('T')


class RingBuffer(Generic[T]):
    """Thread-safe ring buffer with fixed capacity."""
    
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.buffer: List[Optional[T]] = [None] * capacity
        self.head = 0
        self.size = 0
        self.lock = threading.RLock()
    
    def append(self, item: T) -> None:
        """Add item to buffer, overwriting oldest if full."""
        with self.lock:
            self.buffer[self.head] = item
            self.head = (self.head + 1) % self.capacity
            
            if self.size < self.capacity:
                self.size += 1
    
    def get(self, index: int) -> Optional[T]:
        """Get item at index (0 = oldest, -1 = newest)."""
        with self.lock:
            if index >= self.size or index < -self.size:
                return None
            
            if index < 0:
                index = self.size + index
            
            actual_index = (self.head - self.size + index) % self.capacity
            return self.buffer[actual_index]
    
    def get_all(self) -> List[T]:
        """Get all items in chronological order."""
        with self.lock:
            if self.size == 0:
                return []
            
            items = []
            start_index = (self.head - self.size) % self.capacity
            
            for i in range(self.size):
                index = (start_index + i) % self.capacity
                if self.buffer[index] is not None:
                    items.append(self.buffer[index])
            
            return items
    
    def get_latest(self, count: int) -> List[T]:
        """Get latest N items."""
        with self.lock:
            if count <= 0 or self.size == 0:
                return []
            
            count = min(count, self.size)
            items = []
            
            for i in range(count):
                index = (self.head - count + i) % self.capacity
                if self.buffer[index] is not None:
                    items.append(self.buffer[index])
            
            return items
    
    def clear(self) -> None:
        """Clear all items from buffer."""
        with self.lock:
            self.buffer = [None] * self.capacity
            self.head = 0
            self.size = 0
    
    def __len__(self) -> int:
        """Get current size of buffer."""
        with self.lock:
            return self.size
    
    def __bool__(self) -> bool:
        """Check if buffer has items."""
        return len(self) > 0
    
    def is_full(self) -> bool:
        """Check if buffer is full."""
        with self.lock:
            return self.size == self.capacity
    
    def get_capacity(self) -> int:
        """Get buffer capacity."""
        return self.capacity